pkg/provider/doppler/client/client.go

```diff
@@ -0,0 +1,348 @@
+	client := &DopplerClient{
+		DopplerToken: dopplerToken,
+		VerifyTLS:    true,
+		UserAgent:    "doppler-external-secrets",
+	}
+
+	if err := client.SetBaseURL("https://api.doppler.com"); err != nil {
+		return nil, &APIError{Err: err, Message: "setting base URL failed"}
+	}
+
+	return client, nil
+}
+
+func (c *DopplerClient) BaseURL() *url.URL {
+	u := *c.baseURL
+	return &u
+}
+
+func (c *DopplerClient) SetBaseURL(urlStr string) error {
+	baseURL, err := url.Parse(strings.TrimSuffix(urlStr, "/"))
+
+	if err != nil {
+		return err
+	}
+
+	if baseURL.Scheme == "" {
+		baseURL.Scheme = "https"
+	}
+
+	c.baseURL = baseURL
+	return nil
+}
+
+func (c *DopplerClient) Authenticate() error {
+	//  Choose projects as a lightweight endpoint for testing authentication
+	if _, err := c.performRequest("/v3/projects", "GET", headers{}, queryParams{}, httpRequestBody{}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *DopplerClient) GetSecret(request SecretRequest) (*SecretResponse, error) {
+	params := request.buildQueryParams(request.Name)
+	response, err := c.performRequest("/v3/configs/config/secret", "GET", headers{}, params, httpRequestBody{})
+	if err != nil {
+		return nil, err
+	}
+
+	var data secretResponseBody
+	if err := json.Unmarshal(response.Body, &data); err != nil {
+		return nil, &APIError{Err: err, Message: "unable to unmarshal secret payload", Data: string(response.Body)}
+	}
+
+	if data.Value.Computed == nil {
+		return nil, &APIError{Message: fmt.Sprintf("secret '%s' not found", request.Name)}
+	}
+
+	return &SecretResponse{Name: data.Name, Value: *data.Value.Computed}, nil
+}
+
+// GetSecrets should only have an ETag supplied if Secrets are cached as SecretsResponse.Secrets will be nil if 304 (not modified) returned.
+func (c *DopplerClient) GetSecrets(request SecretsRequest) (*SecretsResponse, error) {
+	headers := headers{}
+	if request.ETag != "" {
+		headers["if-none-match"] = request.ETag
+	}
+	if request.Format != "" && request.Format != "json" {
+		headers["accept"] = "text/plain"
+	}
+
+	params := request.buildQueryParams()
+	response, apiErr := c.performRequest("/v3/configs/config/secrets/download", "GET", headers, params, httpRequestBody{})
+	if apiErr != nil {
+		return nil, apiErr
+	}
+
+	if response.HTTPResponse.StatusCode == 304 {
+		return &SecretsResponse{Modified: false, Secrets: nil, ETag: request.ETag}, nil
+	}
+
+	eTag := response.HTTPResponse.Header.Get("etag")
+
+	// Format defeats JSON parsing
+	if request.Format != "" {
+		return &SecretsResponse{Modified: true, Body: response.Body, ETag: eTag}, nil
+	}
+
+	var secrets Secrets
+	if err := json.Unmarshal(response.Body, &secrets); err != nil {
+		return nil, &APIError{Err: err, Message: "unable to unmarshal secrets payload"}
+	}
+	return &SecretsResponse{Modified: true, Secrets: secrets, Body: response.Body, ETag: eTag}, nil
+}
+
+func (c *DopplerClient) UpdateSecrets(request UpdateSecretsRequest) error {
```

Comment:
Required for e2e tests which will be coming in a future commit as they advised us not to add them now.